### ISSUE TITLE
Connect callback is called multiple times if there is a network error during connection

### ISSUE THREAD

**Issue by efarrar** [SAP/node-hdb#64]
During automated testing we have been seeing rare crashes due to what appears to be the callback on connect() being invoked multiple times. These appear during long automated tests during periods of network disruption. I have not been able to build a simple repro. 

When we have network disruptions, usually the error that is returned is:

`Error: Could not connect to any host: [ <<hostname>> - connect ECONNREFUSED <<ip address and port>> ]`

These are handled properly. However, in rare cases when an ECONNRESET is raised during connection, we receive:

`Received connect error: Error: Could not connect to any host: [ <<hostname>> - read ECONNRESET ]`

When the ECONNRESET is received, it appears the callback supplied to connect is being called twice.

I believe this may be because in ConnectionManager.js, a handler is registered for the 'error' event (https://github.com/SAP/node-hdb/blob/master/lib/protocol/ConnectionManager.js#L81). The handler is not removed until after fetchDbConnectInfo has been called (https://github.com/SAP/node-hdb/blob/master/lib/protocol/ConnectionManager.js#L102).

If a network error arises during the fetchDbConnectInfo call, I believe that both the error handler will be called, and fetchDbConnectInfo will return the error, which will cause the callback to be called twice.

There is a specific test case for this: https://github.com/SAP/node-hdb/blob/master/test/hdb.Client.js#L509. However, I believe the mock is not correctly mimicking the connection object. Specifically, this test redefines fetchDbConnectInfo to emit the 'error' (https://github.com/SAP/node-hdb/blob/master/test/hdb.Client.js#L516). However, I believe the real connection object would both emit the error, and call a callback with the error. 

I made the changes on my fork (https://github.com/efarrar/node-hdb/commit/a6b7b4dd2a0b79b743b5ebd2b4177650050c081c), and when I run the test, I see that the callback is being invoked twice. (Unfortunately, mocha does not detect it because the second invocation happens after that test case has passed, so mocha silently ignores it ). In the output, I see: 

```
     should report an error if instance number string is not valid
Connect callback was called
      should report an error if error event has occurred during fetching DB_CONNECT_INFO
Connect callback was called
      should close the connection to system-db in case of error during opening it
```

**Comment by petar-iv** [comment]
Thank you for your analysis. I have prepared a possible fix in [this fork, this branch](https://github.com/petar-iv/node-hdb/tree/fix-calling-callback-twice-2).

I could not reproduce the issue itself on a real system. Thus, I could not ensure that the fix actually solves the problem. Would it be possible to ask you to fetch the hdb driver from the forked repo and the corresponding branch and to try it during your tests to verify the fix? Thanks a lot.

Here is how to fetch it:

```sh
npm install git+https://git@github.com/petar-iv/node-hdb.git#fix-calling-callback-twice-2
```

Let me know what the outcome is.

**Comment by alexpenev-s** [comment]
Should be fixed in v0.10.0
